cmd/test-generation/generate-test.go

```diff
@@ -0,0 +1,342 @@
+// Generates a full netsim test, given a log-splicer processed ssb-fixtures folder and a replication expectations file
+// expectations.json.
+package main
+
+import (
+	"encoding/json"
+	"flag"
+	"fmt"
+	"log"
+	"math/rand"
+	"os"
+	"path"
+	"sort"
+)
+
+func check(err error) {
+	if err != nil {
+		log.Fatalln(err)
+	}
+}
+
+func readExpectations(expectationsPath string) (map[string][]string, error) {
+	b, err := os.ReadFile(expectationsPath)
+	if err != nil {
+		return nil, err
+	}
+	var v map[string][]string
+	err = json.Unmarshal(b, &v)
```

Comment:
try getting used to `json.NewDecoder(exFile).Decode(&v)` with `exFile, err = os.Open(expectationsPath)`.

for small files it's a bit overkill but its better practice for when you try to slurp up very big json content.

(ps `defer exFile.Close()` too)

Reply:
is the point i'm missing that using decoders allows one to decode in a stream, instead of having to buffer the entire contents of a file?

Reply:
right. When you use `dec:=json.NewDecoder(r)` on a stream of `{obj}\n{obj}\n` you can call `err = dec.Decode(&v)` repeatedly to get the single objects.

If you compare these two, you see that the streaming decoder can isolate single values while the other works on the whole thing at once.

https://github.com/golang/go/blob/5c59e11f5e1fe2e6d5b684f9a348022c07807126/src/encoding/json/decode.go#L106
https://github.com/golang/go/blob/5c59e11f5e1fe2e6d5b684f9a348022c07807126/src/encoding/json/stream.go#L67

Again, the actual performance difference only pays of once you slurp really big data and actually process it in chunks compared to marshaling into an array directly.